sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,230 @@
+        else:
+            radius[mask] = 0
+        label_counts[mask] = count
+
+    nn.set_params(algorithm='kd_tree')
+    nn.fit(c)
+    ind = nn.radius_neighbors(radius=radius, return_distance=False)
+    neighbor_counts = np.array([i.size for i in ind])
+
+    mi = (digamma(n_samples) + digamma(n_neighbors) -
+          np.mean(digamma(label_counts)) -
+          np.mean(digamma(neighbor_counts + 1)))
+
+    return max(0, mi)
+
+
+def _compute_mi(x, y, x_categorical, y_categorical, n_neighbors=3):
+    if x_categorical and y_categorical:
+        return mutual_info_score(x, y)
+    elif x_categorical and not y_categorical:
+        return _compute_mi_cd(y, x, n_neighbors)
+    elif not x_categorical and y_categorical:
+        return _compute_mi_cd(x, y, n_neighbors)
+    else:
+        return _compute_mi_cc(x, y, n_neighbors)
+
+
+class MinRedundancyMaxRelevance(BaseEstimator, SelectorMixin):
+    """Select the subset of features based on mutual information criterion.
+
+    The algorithm finds a subset of features jointly having maximum mutual
+    information with the target variable. To avoid estimation of mutual
+    information in high dimensional space the algorithm uses theoretically
+    equivalent incremental search, where the difference between relevance and
+    redundancy of a feature is considered on each iteration [1]_.
+
+    Read more in :ref:`User Guide <MI>`.
+
+    Parameters
+    ----------
+    n_features_to_select : float or int, optional (default=0.5)
+        Number of features to select. The value greater than or equal 1 is
+        interpreted as the absolute number of features to select. The value
+        within (0.0, 1.0) is interpreted as the percentage from the initial
+        number of features (rounded down). Half of the features is selected by
+        default.
+    categorical_features : bool or array_like with shape (n_features),
+                           optional (default=False)
+        If bool, then determines whether to consider all features categorical
+        or continuous. If array, then True elements mark categorical features.
+    categorical_target : bool, optional (default=False)
+        Whether to consider `y` as a categorical variable.
+    n_neighbors : int, optional (default=3)
+        Number of neighbors to use for mutual information estimation for
+        continuous variables, see [2]_ and [3]_. The default value is usually
+        sufficient.
+
+    Attributes
+    ----------
+    support_ : ndarray, shape (n_features,)
+        Mask of selected features.
+    relevance_ : ndarray, shape (n_features,)
+        Array containing mutual information values between each feature and
+        the target variable.
+    redundancy_ : ndarray, shape (n_features, n_features)
+        Array containing mutual information values between each pair of
+        features. Note, that diagonal elements are not computed and contain
+        some random numbers.
```

Comment:
Well technically, the diagonal elements should be the entropy.

We can remove this attribute, if you consider my comment below.